File: 141.linked-list-cycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
	sp, fp := head, head
	for sp != nil && fp != nil {
		if fp.Next != nil {
			fp = fp.Next.Next
		} else {
			return false
		}
		sp = sp.Next
		if fp == sp {
			return true
		}
	}
	return false
}

File: 141.linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        p1, p2 = head, head
        while p1 and p2 and p2.next:
            p1, p2 = p1.next, p2.next.next
            if p1 == p2:
                return True
        return False
